storage/statedb/cache_redis_test.go

```diff
@@ -98,22 +99,66 @@ func TestNewRedisCache(t *testing.T) {
 	assert.Equal(t, bytes.Compare(value, hasValue), 0)
 }
 
-// TestNewRedisCache_Set_LargeData check whether redis cache can store an large data (5MB).
-func TestNewRedisCache_Set_LargeData(t *testing.T) {
+// TestRedisCache_Set_LargeData check whether redis cache can store an large data (5MB).
+func TestRedisCache_Set_LargeData(t *testing.T) {
 	cache, err := NewRedisCache(getTestRedisConfig())
 	if err != nil {
 		t.Fatal(err)
 	}
 
 	key, value := randBytes(32), randBytes(5*1024*1024) // 5MB value
 	cache.Set(key, value)
+	time.Sleep(100 * time.Millisecond)
 
 	retValue := cache.Get(key)
 	assert.Equal(t, bytes.Compare(value, retValue), 0)
 }
 
-// TestNewRedisCache_Timeout test timout feature of redis client.
-func TestNewRedisCache_Timeout(t *testing.T) {
+func TestRedisCache_Set_LargeNumberItems(t *testing.T) {
+	cache, err := NewRedisCache(getTestRedisConfig())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	itemsLen := redisSetItemChannelSize * 2
+	items := make([]setItem, itemsLen)
+	for i := 0; i < itemsLen; i++ {
+		items[i].key = randBytes(32)
+		items[i].value = randBytes(500)
+	}
+
+	// successCh returns struct{} when the following goroutine checks all items are set in the redis
+	successCh := make(chan struct{})
+	go func() {
+		for i := 0; i < itemsLen; i++ {
+			v := cache.Get(items[i].key)
+			if !bytes.Equal(v, items[i].value) {
+				// if the item is not set yet, wait and retry
+				time.Sleep(100 * time.Millisecond)
+				i--
+			}
+		}
```

Comment:
```suggestion
		for i := 0; i < itemsLen; i++ {
			MaxRetry := 5
			v := cache.Get(items[i].key)
			if len(v) == 0 && MaxRetry > 0 {
				// if the item is not set yet, wait and retry
				time.Sleep(100 * time.Millisecond)
				i--
				MaxRetry--
			} else if MaxRetry <= 0 {
				t.failNow()
			} else {
				assert.Equal(v, items[i].value)
			}
		}
```
1. I think the comparing should be done with `assert.Equal`.
2. How about replacing the check condition with `if len(items[i].value)` for retrying logic.
3. How about setting `maxRetry` for each item. If there is a problem connecting to a redis server, this code should wait more than 3min for the test comes out.

Reply:
**Answer for 2, 3**
The main thread waits just 5 seconds for `successCh`. When the "for" loop lasts more than 5 seconds, the test will be finished with `t.Fatal()`.
```
	select {
	case <-successCh:
	case <-time.After(5 * time.Second):
		t.Fatal("timeout")
	}
```

**Answer for 1**
I wanted to make the test code simple, but it would be more precise if the "if" branch in your qute is changed adopting your suggestion.
```
if v == nil {
	time.Sleep(100 * time.Millisecond)
	i--
} else {
	assert.Equal(t, v, items[i].value)
}
```
How do you think about it?

Reply:
FYI, `t.FailNow()` in goroutines cannot finish the test. 
```
// FailNow must be called from the goroutine running the
// test or benchmark function, not from other goroutines
// created during the test. Calling FailNow does not stop
// those other goroutines.
```

Reply:
Thank you for considering.

Reply:
What do you think of switching the go routine part? 'Set` is done in a new go routine, `Get` is done in test goroutine. Because the `Get` part has the checking logics, it should be in the main routine.

Reply:
I switched it. Thanks